// @glueit/back/skel/back/default/steps.js
module.exports = [
  {
    dateCreated: '2019-12-10',
    actions: [
      {
        action: 'dropSpace',
        name: 'learn'
      },
      {
        action: 'createSpace',
        name: 'learn'
      },
      {
        action: 'useSpace',
        name: 'learn'
      }
    ]
  },
  {
    dateCreated: '2019-12-10',
    actions: [
      {
        action: 'createObject',
        name: 'posts',
        properties: [
          {
            name: 'id',
            type: 'number',
            on: 'posts'
          },
          {
            name: 'userId',
            type: 'number',
            on: 'posts'
          }
        ]
      },
      {
        action: 'createProperty',
        name: 'title',
        type: 'string',
        length: 64,
        on: 'posts'
      },
      {
        action: 'createProperty',
        name: 'body',
        type: 'string',
        length: 256,
        on: 'posts'
      },
      {
        action: 'createObject',
        name: 'users',
        properties: [
          {
            action: 'createProperty',
            name: 'id',
            type: 'number',
            on: 'users'
          }
        ]
      },
      {
        action: 'createProperty',
        name: 'username',
        type: 'string',
        length: 16,
        on: 'users'
      },
      {
        action: 'createProperty',
        name: 'firstName',
        type: 'string',
        length: 32,
        on: 'users'
      },
      {
        action: 'createProperty',
        name: 'lastName',
        type: 'string',
        length: 32,
        on: 'users'
      },
      {
        action: 'createProperty',
        name: 'email',
        type: 'string',
        length: 128,
        on: 'users'
      },
      {
        action: 'createProperty',
        name: 'password',
        type: 'string',
        length: 128,
        on: 'users'
      }
    ]
  },
  {
    dateCreated: '2019-12-11',
    actions: [
      {
        action: 'updateProperty',
        name: 'title',
        type: 'string',
        length: 128,
        on: 'posts'
      }
    ]
  },
  {
    dateCreated: '2019-12-11',
    actions: [
      {
        action: 'updateProperty',
        name: 'title',
        type: 'string',
        length: 256,
        on: 'posts'
      }
    ]
  }
]
